[ZOOKEEPER-3057] Fix IPv6 literal usage

mjeelanimsft opened the pull request:
This patch contains fixes for IPv6 literal usage and corresponding unit test changes. 

As per discussion in ZOOKEEPER-3057 - The issue/problem is the same as ZOOKEEPER-2989, but we changed the code to pass IPv6 literal [%s]:%s, also we changed the logging and the LocalPeerBean to show this IPv6 literal as well, which makes it easier to check when using Ipv6 and we added detailed tests for this change, sending out for review to see if it's better or not.

ZKPatch: 88e94e6f3665353446bf70a042c8f0cd50834f7c (extract)

maoling commented on src/java/main/org/apache/zookeeper/client/FourLetterWordMain.java:
```diff
@@ -86,7 +86,7 @@ public static String send4LetterWord(String host, int port, String cmd, boolean
         LOG.info("connecting to {} {}", host, port);
         Socket sock;
-        InetSocketAddress hostaddress= host != null ? new InetSocketAddress(host, port) :
+        InetSocketAddress hostaddress = host != null ? new InetSocketAddress(host, port) :
```
it seems that this PR includes some code formats and typos which isn't related.

anmolnar replied:
Agreed, it's not ideal, but size of the patch makes it easy to deal with I would say.

lvfangmin replied:
This is generally due to some auto format done in editor.

lvfangmin replied:
@mjeelanimsft generally, this is due to the auto-format in the editor, but I think we should remove this change from this diff, given we haven't changed other logic except this one.

maoling commented on src/java/main/org/apache/zookeeper/server/util/ConfigUtils.java:
```diff
@@ -60,4 +60,25 @@ static public String getClientConfigStr(String configData) {
+    public static String[] splitServerConfig(String s)
+        throws ConfigException
+    {
+        /* Does it start with an IPv6 literal? */
```
good method.this annotation's answer is sure(grin)

maoling replied:
remove this line annotation and give this method detailed javadoc

maoling commented on src/java/test/org/apache/zookeeper/common/NetUtilsTest.java:
```diff
@@ -0,0 +1,46 @@
+package apache.zookeeper.common;
```
???

breed replied:
@mjeelanimsft i think this is why the tests are failing. can you fix this package name? (good catch @maoling !)

mjeelanimsft replied:
Thanks @maoling @breed - working on fixing this

maoling commented:
- It seems that zk server side supports ipv6 style like this: `server.1=[2001:db8:1::242:ac11:2]:2888:3888`,but zk client side supports ipv6 like this:`2001:db8:1::242:ac11:2:2181`.we need unify them?[ConnectStringParser](https://github.com/apache/zookeeper/blob/master/src/java/main/org/apache/zookeeper/client/ConnectStringParser.java#L71) can reuse that method?
- IMO,we should document this feature to let user know.(e.g. [here](https://github.com/apache/zookeeper/blob/master/src/docs/src/documentation/content/xdocs/zookeeperAdmin.xml#L1055))

anmolnar commented on src/java/test/org/apache/zookeeper/common/NetUtilsTest.java:
```diff
@@ -0,0 +1,46 @@
+    private String v6local = v6addr + ":" + port.toString();
+
+    @Test
+    public void testformatInetAddrGoodIpv4() {
```
Typo: capital `F`

anmolnar commented on src/java/test/org/apache/zookeeper/common/NetUtilsTest.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    @Test
+    public void testFormatInetAddrGoodIpv6() {
```
I think it'd be great to add more IPv6 formatting tests other than the localhost case.

mjeelanimsft replied:
Thanks @anmolnar - I'll add the additional tests

lvfangmin commented on src/java/main/org/apache/zookeeper/server/quorum/QuorumCnxManager.java:
```diff
@@ -225,17 +229,25 @@ static public InitialMessage parse(Long protocolVersion, DataInputStream din)
             if (host_port.length != 2) {
                 throw new InitialMessageException("Badly formed address: %s", addr);
             }
+            //String[] host_port = addr.split(":");
```
Remove this line.

lvfangmin commented on src/java/test/org/apache/zookeeper/server/quorum/ReconfigFailureCasesTest.java:
```diff
@@ -58,10 +58,10 @@ public void tearDown() throws Exception {
 
     /*
-     * Tests that an incremental reconfig fails if the current config is hiearchical.
+     * Tests that an incremental reconfig fails if the current config is hierarchical.
```
Please remove this unrelated change as well.

anmolnar commented:
@mjeelanimsft Just to clarify my latest comment: I'd to see a new class `ConfigUtilsTest` in which you explicitly verify the behaviour of `splitServerConfig`. Thanks.

mjeelanimsft commented:
Thanks @anmolnar & @maoling - I've added the Unit Test as well as the JavaDoc for splitServerConfig() method

anmolnar commented:
@mjeelanimsft Sorry, I still can't see unit tests for `splitServerConfig`.

anmolnar commented:
@mjeelanimsft Now I can see it, thanks.
I'd write 3 tests at the minimum to validate all if branches in the method:
With and w/o the `[` prefix and a third one for the invalid case: no matching closing brace.

mjeelanimsft commented:
Thanks Andor - I've added the unit test cases - For the one w/o [ - its also an invalid case for this method as its needed to specify the port. I've added that as one at the end, let me know what you think. 

1) Successful case (IPv6)
2) Successful case (IPv4)
3) Invalid case (IPv6 with missing closing brackets)
4) Invalid case (IPv6 with port but missing brackets)

anmolnar commented:
@mjeelanimsft  I keep trying to refresh the page, but again I can't see the new tests.
Also the patch has some conflicts with the latest master, would you please resolve them?

mjeelanimsft commented:
@anmolnar Resolved the merge conflict. You can see the commit with the unit test changes here - https://github.com/apache/zookeeper/pull/548/commits/ff4704984203921071877bdee4efef47b8c2a706

maoling commented on src/java/main/org/apache/zookeeper/server/util/ConfigUtils.java:
```diff
@@ -61,10 +61,16 @@ static public String getClientConfigStr(String configData) {
+     * @param s server config, server:port
+     */
     public static String[] splitServerConfig(String s)
         throws ConfigException
```
`splitServerConfig` is somewhat general. `getHostAndPort` is better?

mjeelanimsft replied:
Sure, I like that better too - I've renamed this

maoling commented on src/java/main/org/apache/zookeeper/server/quorum/QuorumCnxManager.java:
```diff
@@ -226,17 +230,25 @@ static public InitialMessage parse(Long protocolVersion, DataInputStream din)
             // FIXME: IPv6 is not supported. Using something like Guava's HostAndPort
             //        parser would be good.
```
remove this ?

maoling commented on src/java/main/org/apache/zookeeper/server/quorum/QuorumCnxManager.java:
```diff
@@ -857,15 +869,15 @@ public void run() {
                     }
-                    LOG.info("My election bind port: " + addr.toString());
+                    LOG.info("My election bind port: " + formatInetAddr(addr));
                     setName(addr.toString());
```
using  `toString` can still distinguish ipv4 from ipv6 by looking at the ip style although without `[] ` 
if `formatInetAddr(addr)` is really needed, can you make sure all the places which log `InetSocketAddress` are covered?

mjeelanimsft replied:
Good point - I'll remove formatInetAddr(addr) from the log

mjeelanimsft commented:
Thanks for the feedback @maoling - I've incorporated the below changes. @maoling @anmolnar - let me know what you think

- Assert actual host and port in unit tests for getHostAndPort
- Renamed splitServerConfig(String s) to getHostAndPort(String s)
- Removed Fix Me comment in QuorumCnxManager.java
- No need to formatInetAddr(addr) in QuorumCnxManager.java log.Info - removed it

breed commented:
+1 i'll kick jenkins

maoling commented:
ping @mjeelanimsft 
Would you please create separate pull request for branch-3.5 and branch-3.4?

xy24 commented:
@anmolnar @maoling 
I've cherry picked this change and created a pr for branch-3.5.
Would you please review it for me? Thanks
#1168
